alloc-fmt/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+// Copyright 2017 the authors. See the 'Copyright and license' section of the
+// README.md file at the top-level directory of this repository.
+//
+// Licensed under the Apache License, Version 2.0 (the LICENSE-APACHE file) or
+// the MIT license (the LICENSE-MIT file) at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Allocator-safe formatting and assertion macros.
+//!
+//! `alloc-fmt` provides formatting and assertion macros similar to the standard library's
+//! `println`, `eprintln`, `assert`, `debug_assert`, etc which are safe for use in an allocator.
+//! The standard library's formatting and assertion macros can allocate, meaning that if they are
+//! used in the implementation of an allocator, it can cause infinite recursion. The macros in this
+//! crate avoid this problem by either not allocating (in the case of formatting macros) or
+//! detecting recursion (in the case of assertion macros).
+//!
+//! # Usage and Behavior
+//! The macros in this crate are named `alloc_xxx`, where `xxx` is the name of the equivalent
+//! standard library macro (e.g., `alloc_println`, `alloc_debug_assert`, etc).
+//!
+//! The behavior of the formatting macros is identical to the behavior of their standard library
+//! counterparts. The behavior of the assertion macros is somewhat different. When an assertion
+//! fails, a message is first unconditionally printed to stderr (in case further processing causes
+//! a crash). A stack trace is then printed, and the process aborts. If, during the printing of the
+//! stack trace, recursion is detected (computing stack traces can cause allocation, which can
```

Comment:
The check for recursion is conservative, no? 

Maybe reword as "If possible recursion is detected while printing the stack trace (e.g. if creating the stack trace results in a call to `alloc` which in turn panics ... ) the process aborts immediately"